datastreams/servicename.go

```diff
@@ -0,0 +1,45 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package datastreams
+
+import "strings"
+
+// NormalizeServiceName translates serviceName into a string suitable
+// for inclusion in a data stream name.
+//
+// Concretely, this function will lowercase the string and replace any
+// reserved characters with "_".
+func NormalizeServiceName(s string) string {
+	s = strings.ToLower(s)
+	s = strings.Map(replaceReservedRune, s)
```

Comment:
There are consequences to this, I think if it a replacement actually happens we should log a warn. And maybe create a doc issue / note specifically for this?

Reply:
I expect it will be pretty common, given that people can and do use hyphens in their service names. Logging warnings would suggest that they should not hyphenate, which is unreasonable.

The UI is probably a more appropriate place to call attention to this anyway, rather than expecting people to go digging in logs. We can do that by looking for reserved characters in `service.name` values if necessary.

We should certainly document the behaviour. I've added a TODO list item to https://github.com/elastic/apm-server/issues/4378.